### PR TITLE
Test Suite Pruning

### PR DESCRIPTION
**Pull request by righthandabacus**
## 📌 Description

This change introduces regular and full modes for large Cartesian pytest parameter matrices:

- A normal `pytest` invocation runs a reduced, deterministic selection of cases.
- Passing `--full` restores the complete Cartesian parameter matrices.
- The nightly test script passes `--full`, so nightly coverage remains complete.

The implementation adds a shared `parametrize_product` helper and registers the `--full` pytest option through the existing test configuration. The helper also supports grouped parameters, such as `(num_q_heads, num_kv_heads)`, that must remain coupled rather than being expanded as independent axes.

The following multi-axis Cartesian tests use the regular/full mechanism, selected based on the timing data which took the most time in the test suite:

- `tests/attention/test_xqa_batch_decode.py::test_xqa_batch_decode`
- `tests/attention/test_trtllm_gen_attention_decode_xqa.py::test_trtllm_batch_decode`
- `tests/utils/test_sampling.py::test_top_k_top_p_sampling_from_probs_logits_alignment`
- `tests/attention/test_xqa.py::test_xqa`
- `tests/gemm/test_fp8_blockscale_gemm.py::test_fp8_blockscale_gemm_sm90`
- `tests/gdn/test_decode_delta_rule.py::test_gdn_decode_fp32_state_fla_scatter_vs_dense`
- `tests/attention/test_attention_sink.py::test_attention_sink`
- `tests/attention/test_deepseek_mla.py::test_batch_mla_varlen_page_attention`
- `tests/attention/test_rope.py::test_generalized_rope_quantize_append_kv_cache`
- `tests/attention/test_batch_attention.py::test_batch_attention_correctness`
- `tests/gdn/test_decode_delta_rule.py::test_gdn_decode_bf16_state_fused_recovery_decode`
- `tests/attention/test_logits_cap.py::test_single_decode_logits_soft_cap`
- `tests/attention/test_batch_decode_kernels.py::test_batch_decode_with_paged_kv_cache`

Amongst these, a total of 615 test nodes are assigned to regular and the full version are 170313 nodes.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

The regular cases are selected independently for each test using deterministic pairwise coverage (a strength-2 covering array). They are not sampled globally from the complete set of test nodes.

A full Cartesian product covers every higher-order interaction across all parameter axes, but grows exponentially. Selecting only one occurrence of every individual value is much cheaper, but can leave parameter values correlated and miss important two-parameter interactions. Pairwise coverage is the compromise:

> For every pair of parameter axes, every combination of their values appears in at least one regular case.

For example, three binary axes have eight full combinations. The following four cases cover every pair of values across `A/B`, `A/C`, and `B/C`:

```text
A0 B0 C0
A0 B1 C1
A1 B0 C1
A1 B1 C0
```

The selector:

1. Builds the set of all required value-pairs across every pair of axes.
2. Generates deterministic candidate rows containing each required pair.
3. Fills the remaining positions deterministically without materializing the complete Cartesian product.
4. Repeatedly selects the candidate covering the most uncovered pairs.
5. Stops when every required pair has appeared.

This greedy algorithm is deterministic and reproducible, although it does not guarantee the mathematically smallest possible covering array. Grouped parameter values remain atomic, preserving the valid combinations from the original tests.

## Summary by CodeRabbit

* **Testing**
  * Regular test runs now use pairwise parameter coverage to reduce redundant combinations.
  * Added a `--full` option for complete parameter matrices; nightly builds use it automatically.
  * Improved test-runner management, parallel compilation limits, and resource monitoring.
  * Slow test collection can now capture diagnostic stack information.

* **Documentation**
  * Updated testing guidance to distinguish regular and full test commands.

* **Chores**
  * H100 test artifacts now include collection diagnostics alongside existing test reports.

### CODE REVIEW

**Comment by aleozlx** [comment]
## PR Review Screening

**CI verdict:** ✅ auto-run ok
**Review category:** `live` (rule fired: C3.2 — a new repo-wide test-parametrization abstraction, registered globally and documented in `CLAUDE.md`, changes the coverage contract for the whole suite)
**Blocking checks:** none
**Release blocker:** no
**Early stop:** no

### Security
| Q | Answer | Evidence |
|---|--------|----------|
| S1 injection/supply-chain | no | — |
| S2 template overwritten | no | — |
| S3 template obligations | met | all five items ticked, Experimental subsection correctly left untouched, and the Reviewer Notes name the technique (strength-2 covering array) and argue the tradeoff explicitly |
| S4 agent-directing text | no | — |

### Packaging
| Q | Answer | Evidence |
|---|--------|----------|
| C1.1 dependency bump | no | no dependency, pin, index or submodule change |
| C1.2 public API changes | no | `run_ninja` gains an optional `max_jobs: Optional[int] = None`; it is an internal JIT helper and the addition is backward compatible. No exported surface changes |
| C1.3 AOT/trace registration | no | `aot.py` untouched; no module generators |

### Presentation
| Q | Answer | Evidence |
|---|--------|----------|
| C2.1 perf claim backed | n-a | no library performance claim — the quantified claim is a test-node count, not kernel throughput |

### Implementation
| Q | Answer | Evidence |
|---|--------|----------|
| C3.1 experimental declared | no | — |
| C3.2 durable areas touched | yes — durable | `tests/test_helpers/parametrize.py` registered as a global plugin via `pytest_plugins` in `tests/conftest.py`, JIT worker-count threading through `flashinfer/jit/core.py` and `cpp_ext.py`, sharding-worker changes, and the convention recorded in `CLAUDE.md`. This is a shared abstraction and a design decision embedded in code, which is what routes it `live` |
| C3.3 tests match change | ❗ no — *executes but does not cover* | `pairwise_product_cases` is ~60 lines of greedy set-cover over a candidate pool, and it now decides which cases 52 test functions run. Nothing asserts its covering property, so an under-selection bug would silently shrink coverage everywhere while every suite stays green. `tests/test_helpers/` is in `norecursedirs`, so such a test needs to live elsewhere (e.g. `tests/utils/`) |
| C3.4 release-blocker candidate | no | test-infrastructure change; no hang, crash, corruption or prior-release regression |

### Experimental track
| Q | Answer | Evidence |
|---|--------|----------|
| C4.1 declaration obligations | n-a | C3.1 = no |
| C4.2 machine-readable test scope | n-a | C3.1 = no |
| C4.3 isolated from common areas | n-a | C3.1 = no |

**Notes for the maintainer**

- **The coverage contract changes shape, and the remainder lands on nightly.** Pairwise keeps every single-parameter and every two-parameter interaction; combinations that require three or more specific axis values now run only under `--full`, which means only in nightly, which means after merge. That is a legitimate and well-understood trade, but it is a decision worth making out loud rather than inheriting — particularly because nightly completeness is itself a tracked concern, so the safety net for higher-order interactions is the lane least guaranteed to finish. The other half of the trade is unstated: the 70% node reduction is reported, but no before/after wall-clock for the lanes this is meant to speed up, and that is the number the change is actually buying.
- **51 of the 52 conversions carry the mechanical guarantee; one does not.** `tests/utils/test_topk.py::test_top_k` uses a hand-authored `regular=[...]` list rather than `pairwise_product_cases`. The list looks deliberately constructed, but it has no all-pairs property by construction and will not preserve one if someone later edits an axis. Worth either converting it or recording why it is hand-picked.
- **Two practical collisions.** `tests/utils/test_topk.py` is also rewritten by #5187 (+255 lines), which is a release-blocker candidate — whichever lands second will conflict there, and it is worth deciding the order deliberately. Separately, converted tests now carry generated `case-{index}` identifiers, so any pinned node ID — a targeted CI path, a `--deselect`, a scoped rerun someone has saved — that names the old descriptive parametrize IDs will stop matching rather than fail loudly.

<sub>Generated by flashinfer-pr-screen · rubric: docs/code_review_guidance.md · not a code review · AI screening can make mistakes — a maintainer's judgment supersedes this report.</sub>
